crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -188,40 +214,354 @@ pub struct StateTransitionEvent<S: States> {
-                    from: exited,
-                    to: entered.clone(),
-                })
-                .ok();
-            world.try_run_schedule(OnEnter(entered)).ok();
+    let Some(next_state_resource) = world.get_resource::<NextState<S>>() else {
+        return;
+    };
+
+    match next_state_resource {
+        NextState::Set(entered) => {
+            let entered = entered.clone();
+            match world.get_resource_mut::<State<S>>() {
+                Some(mut state_resource) => {
+                    if *state_resource != entered {
+                        let exited = mem::replace(&mut state_resource.0, entered.clone());
+                        world.send_event(StateTransitionEvent {
+                            before: exited.clone(),
+                            after: entered.clone(),
+                        });
+                        // Try to run the schedules if they exist.
+                        world.try_run_schedule(OnExit(exited.clone())).ok();
+                        world
+                            .try_run_schedule(OnTransition {
+                                from: exited,
+                                to: entered.clone(),
+                            })
+                            .ok();
+                        world.try_run_schedule(DeriveStates::<S>::default()).ok();
+                        world.try_run_schedule(OnEnter(entered)).ok();
+                    }
+                }
+                None => {
+                    world.insert_resource(State(entered.clone()));
+                    world.try_run_schedule(DeriveStates::<S>::default()).ok();
+                    world.try_run_schedule(OnEnter(entered)).ok();
+                }
+            };
+        }
+        NextState::Remove => {
+            if let Some(resource) = world.remove_resource::<State<S>>() {
+                world.try_run_schedule(OnExit(resource.0)).ok();
+                world.try_run_schedule(DeriveStates::<S>::default()).ok();
+            }
+        }
+        _ => {
+            return;
+        }
+    }
+
+    world.insert_resource(NextState::<S>::Unchanged);
+}
+
+/// Trait defining a derived state
+///
+/// A derived state is a state implementing [`States`] that is deterministically determined from one or more other [`States`].
+pub trait DerivedStates: States {
+    /// The set of states from which the [`Self`] is derived.
+    type SourceStates: StateSet;
+
+    /// A function for deriving the states.
+    fn derive(
+        sources: <<Self as DerivedStates>::SourceStates as StateSet>::Optionals,
+    ) -> Option<Self>;
+
+    /// A function used to register the derived state with the app schedule.
+    fn add_derivation_to_schedule(schedules: &mut Schedules) {
+        Self::SourceStates::add_derivation_systems_for_derived_state::<Self>(schedules);
+    }
+}
+
+mod sealed {
+    pub trait StateSetSealed {}
+}
+
+/// Trait defining valid sets of states used as a source for a Derived State.
```

Comment:
More context on the machinery would be good here.

These docs should also explicitly call out that a) it's a sealed trait so you can't implement it. b) It's blanket-implemented for all state types.